In [15]:
import pandas as pd
cuisines_df = pd.read_csv("cleaned_cuisines.csv")
cuisines_df.shape

(3995, 382)

In [16]:
# import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [18]:
# Prepare the label and feature dataframes
# Features and target
y = cuisines_df['cuisine']
X = cuisines_df.drop(columns=['Unnamed: 0', 'cuisine'], errors='ignore')
# Encode target labels
le = LabelEncoder()
y_encoded = le.fit_transform(y)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y_encoded, test_size=0.2, random_state=42, stratify=y_encoded
)

In [21]:
#1. lets perform Logistic Regression with lbfgs solver
logreg_lbfgs = LogisticRegression(max_iter=500, solver="lbfgs", multi_class="multinomial")
logreg_lbfgs.fit(X_train, y_train)
y_pred_lbfgs = logreg_lbfgs.predict(X_test)
acc_lbfgs = accuracy_score(y_test, y_pred_lbfgs)

c:\Users\LENOVO\miniconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1272: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


In [22]:
# Now lets see the performance of logistic saga solver
logreg_saga = LogisticRegression(max_iter=500, solver="saga", multi_class="multinomial")
logreg_saga.fit(X_train, y_train)
y_pred_saga = logreg_saga.predict(X_test)
acc_saga = accuracy_score(y_test, y_pred_saga)

c:\Users\LENOVO\miniconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1272: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


In [24]:
# get accuracy scores for these solvers
print("lbfgs accuracy score: {:.2f}".format(acc_lbfgs))
print("saga accuracy score: {:.2f}".format(acc_saga))

lbfgs accuracy score: 0.83
saga accuracy score: 0.82


Although both have good scores, 'lbfgs accuracy score' stands out (0.83).
If your dataset is small to medium, dense (like cuisines/ingredients), lbfgs is faster and more stable.

If your dataset is large and sparse (e.g., millions of text features in NLP, bag-of-words, TF-IDF) saga is the solver of choice.